### DISCUSSION TITLE
64-bit variable types support for ANE

### DISCUSSION THREAD

**Discussion by jovanpn** [airsdk/Adobe-Runtime-Support#2319] [Q&A]
Hi,

I have an old extension (not created by me) which I use to load .net dll assemblies and then run .net code from ActionScript. Now I'm building an AIR application which should be compiled as Windows 64-bit application but I'm having issues with this extension.
Extension is written in C++ and I have its source code so I wanted to update it to support 64-bit applications, but as mentioned above I'm having some issues. Extension code is creating .net objects after they are ported from ActionScript to ExtensionContext and it is using pointers to manage objects.
The problem I have is that extension uses FREGetObjectAsInt32 function from FlashRuntimeExtensions library to read pointers from the object and FRENewObjectFromInt32 to store them into the objects and this is working fine in 32-bit application, but in 64-bit application, pointers are 64-bit integers by default and then in extension they are being truncated to 32-bit values when they are written into the object. This at the end causes objects to be lost in memory.
I've tried using FREGetObjectAsDouble function to save pointers but the effect is more or less the same. So I guess that if some kind of function like FREGetObjectAsInt64 for 64-bit applications would exist in FlashRuntimeExtensions library, this would work for me.
Does anyone have any other suggestion how I could solve this problem and does anybody knows are there any plans on adding this 64-bit integer types support?

Thanks in advance.

**Comment by ajwfrost** [comment]
Hi

I'm slightly surprised that the FREGetObjectAsDouble methods didn't work if you were casting it properly, doubles are just stored internally as 64-bit values into memory.

But the alternative at the moment would be to split it into two 32-bit values and assign it to two 'FREObject' handles, e.g. using the Win32 `LARGE_INTEGER` structure (https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-large_integer-r1)

We are planning to add native support for 64-bit integers in the future but it's not there yet I'm afraid!

thanks

**Comment by jovanpn** [comment]
Hi ajwfrost,

thank you very much for your answer and your suggestions and sorry for my late response. I took some time to properly investigate this problem and try your suggestion. I've come up with most likely reasons why FREGetObjectAsDouble methods doesn't work and I'll try to explain it.
The extension I'm using has ability to load .net assembly files (.dll) dynamically during runtime and then developer can use those .net functions from ActionScript. And all of my investigations went around this function because this is the place where the problem happens.
This is content of a function (with some of mine lines written for investigation purposes):


	FREObject load_assembly_file(FREContext ctx, void* funcData, uint32_t argc, FREObject argv[])
	{
		//----------------------------
		//Get File Path
		const uint8_t* str_path = NULL;
		uint32_t str_path_lenght = 0;
		FREGetObjectAsUTF8(argv[0], &str_path_lenght, &str_path);

		//----------------------------
		//Creating Handle
		FREObject toresult;
		DotNetObject* assembly = new DotNetObject(Assembly::LoadFrom(UTF8toString(str_path)), ctx);

		char temp_buffer[64];
		sprintf(temp_buffer, "%p", assembly);
		MessageBox::Show("Assembly pointer: " + gcnew String(temp_buffer));
		sprintf(temp_buffer, "%p", static_cast<double>((int64_t)assembly));
		MessageBox::Show("Assembly static cast to double: " + gcnew String(temp_buffer));
		MessageBox::Show("Assembly variable type: " + gcnew String(typeid(assembly).name()));
		double assemblyD = (int64_t)assembly;
		MessageBox::Show("Assembly casted to double: " + assemblyD);
		MessageBox::Show("Assembly pointer and its double equal: " + ((int64_t)assembly == assemblyD));

	#ifdef _WIN64
		FRENewObjectFromDouble((int64_t)assembly, &toresult);
	#else
		FRENewObjectFromInt32((int32_t)assembly, &toresult);
	#endif

		DotNetObject* dno;
	#ifdef _WIN64
		FREGetObjectAsDouble(toresult, (double*)&dno);
	#else
		FREGetObjectAsInt32(toresult, (int32_t*)&dnoD);
	#endif

		MessageBox::Show("DNO variable type: " + gcnew String(typeid(dno).name()));
		sprintf(temp_buffer, "%p", dno);
		MessageBox::Show("DNO pointer: " + gcnew String(temp_buffer));

		return toresult;
	}

The problem starts even inside this same function where double object is created. I can confirm that FRENewObjectFromDouble doesn't make the problem because this double number is correctly read within ActionScript code.
The problem is in FREGetObjectAsDouble method. There the double number is wrongly casted. The value of dno pointer is the same as static_cast<double>((int64_t)assembly) from the line above, but this isn't the actual assembly pointer value which then leads extension to read assembly object from wrong memory location. The same situation appears in any other extension method when code tries to read the same double object. In 32bit application this works perfectly because extension uses FREGetObjectAsInt32 method.
So from my perspective, this looks to me more like AIR extension code problem than C++.

I then wanted to try your suggestion to use LARGE_INTEGER structure and I've come up with some solution through ActionScript, although I had to avoid using bitwise right shift operator because it converts number and shift count number to 32bit integer first.
So at the end it seems we have some solution while we wait for full 64bit data types support from AIR.

Thank you again for your help and support!

**Reply by ajwfrost** [reply]
Okay thanks - I've not tried it myself but I would hope that storing a double value, via `FRENewObjectFromDouble`, and then retrieving it, via `FREGetObjectAsDouble`, would always work i.e. you get what you had set.  So if that isn't the case, we will need to investigate!

But if it is the case then it may be a little simpler to work with a union so that the types are a little clearer. e.g.
```
union {
  DotNetObject* pVal;
  double dVal;
}
```
and then maybe just check that the size of that structure is 8 bytes (which should be the case on both 32-bit and 64-bit platforms), then you can set/get the AS3 value using "dVal" but the "pVal" should be correctly set as a pointer to your native object.

Anyway -> we took a brief look at the idea of getting 64-bit integer support into AIR and avmplus .. it's going to be a complex one I think!

thanks
